libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/voice/speech/InstructionDownloadTask.java

```diff
@@ -0,0 +1,82 @@
+package com.mapbox.services.android.navigation.ui.v5.voice.speech;
+
+import android.os.AsyncTask;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import okhttp3.ResponseBody;
+
+class InstructionDownloadTask extends AsyncTask<ResponseBody, Void, File> {
+  private static int instructionNamingInt = 1;
+  private final String cacheDirectory;
+  private final TaskListener taskListener;
+
+  InstructionDownloadTask(String cacheDirectory, TaskListener taskListener) {
+    this.cacheDirectory = cacheDirectory;
+    this.taskListener = taskListener;
+  }
+
+  @Override
+  protected File doInBackground(ResponseBody... responseBodies) {
+    return saveAsFile(responseBodies[0]);
+  }
+
+  /**
+   * Saves the file returned in the response body as a file in the cache directory
+   *
+   * @param responseBody containing file
+   * @return resulting file, or null if there were any IO exceptions
+   */
+  private File saveAsFile(ResponseBody responseBody) {
```

Comment:
https://github.com/mapbox/mapbox-navigation-android/pull/751#discussion_r179169728
I see three clear steps going on here (also you're denoting them adding new lines between them):
1. Arrange everything needed
2. Retrieve/store the file
3. Return the file